src/core/algorithms/sfd/cords/cords.cpp

```diff
@@ -0,0 +1,289 @@
+            zeros_sum += IsZero(n_i_j[i][j]);
+        }
+    }
+    return zeros_sum > min_structural_zeroes_amount_ * domains[col_i] * domains[col_k];
+}
+
+/* Similar to formulae (1) from "CORDS: Automatic Discovery of Correlations and Soft
+   Functional Dependencies". Seems like formulae which is given in the paper is
+   incorrect. */
+long double Cords::ChiSquared(std::vector<std::vector<long double>> const &n_i_j,
+                              std::vector<long double> const &n_i,
+                              std::vector<long double> const &n_j, size_t col_i, size_t col_k,
+                              long double sample_size) {
+    long double chi_squared = 0;
+    for (int i = 0; i < domains[col_i]; i++) {
+        for (int j = 0; j < domains[col_k]; j++) {
+            if (n_i[i] * n_j[j] == 0) return 0;
+            long double actual = n_i_j[i][j];
+            long double expected = n_i[i] * n_j[j] / sample_size;
+            chi_squared += (actual - expected) * (actual - expected) / (expected);
+        }
+    }
+    return chi_squared;
+}
+
+void Cords::Init(size_t columns) {
+    is_skewed.resize(columns, false);
+    domains.resize(columns, 0);
+}
+
+std::pair<size_t, size_t> Cords::SwitchIndicesIfNecessary(size_t ind1, size_t ind2, size_t card1,
+                                                          size_t card2) {
+    if (card2 > card1) {
+        return {ind2, ind1};
+    }
+    return {ind1, ind2};
+}
+
+unsigned long long Cords::ExecuteInternal() {
+    std::vector<model::TypedColumnData> const &data = typed_relation_->GetColumnData();
+
+    size_t rows = data[0].GetNumRows();
+    size_t columns = data.size();
+    std::vector<bool> is_soft_or_trivial(columns, false);
+    Init(columns);
+    auto start_time = std::chrono::high_resolution_clock::now();
+    FrequencyHandler handler(data, columns, max_amount_of_categories_);
+
+    for (size_t ind1 = 0; ind1 < columns - 1; ind1++) {
+        if (is_soft_or_trivial[ind1]) {
+            continue;
+        }
+        if (handler.GetColumnCardinality(ind1) >= (1 - minimum_cardinalty_) * rows) {
+            soft_keys_.push_back(ind1);
+            is_soft_or_trivial[ind1] = true;
+            continue;
+        }
+        if (handler.GetColumnCardinality(ind1) == 1) {
+            trivial_columns_.push_back(ind1);
+            is_soft_or_trivial[ind1] = true;
+            continue;
+        }
+
+        for (size_t ind2 = ind1 + 1; ind2 < columns; ind2++) {
+            if (is_soft_or_trivial[ind2]) {
+                continue;
+            }
+            if (handler.GetColumnCardinality(ind2) >= (1 - minimum_cardinalty_) * rows) {
+                soft_keys_.push_back(ind2);
+                is_soft_or_trivial[ind2] = true;
+                continue;
+            }
+            if (handler.GetColumnCardinality(ind2) == 1) {
+                trivial_columns_.push_back(ind2);
+                is_soft_or_trivial[ind1] = true;
+                continue;
+            }
```

Comment:
try to avoid code duplication from 198 to 210 lines (extract to separate function/lambda)